File: src/main/java/com/leijendary/spring/microservicetemplate/mapper/DateMapper.java
package com.leijendary.spring.microservicetemplate.mapper;

import org.mapstruct.Mapper;

import java.time.OffsetDateTime;

import static com.leijendary.spring.microservicetemplate.util.RequestContext.getTimeZone;
import static java.time.Instant.ofEpochMilli;
import static java.time.OffsetDateTime.ofInstant;
import static org.mapstruct.factory.Mappers.getMapper;

@Mapper
public interface DateMapper {

    DateMapper INSTANCE = getMapper(DateMapper.class);

    default OffsetDateTime toOffsetDateTime(final long epochMillis) {
        final var instant = ofEpochMilli(epochMillis);
        final var zoneId = getTimeZone().toZoneId();

        return ofInstant(instant, zoneId);
    }

    default long toEpochMillis(final OffsetDateTime offsetDateTime) {
        return offsetDateTime.toInstant().toEpochMilli();
    }
}

File: src/main/kotlin/com/leijendary/spring/core/mapper/DateMapper.kt
package com.leijendary.spring.core.mapper

import com.leijendary.spring.core.util.RequestContext.zoneId
import org.mapstruct.Mapper
import org.mapstruct.factory.Mappers.getMapper
import java.sql.Timestamp
import java.time.Instant
import java.time.OffsetDateTime

@Mapper
interface DateMapper {
    companion object {
        val INSTANCE: DateMapper = getMapper(DateMapper::class.java)
    }

    fun toOffsetDateTime(epochMilli: Long): OffsetDateTime {
        return Instant
            .ofEpochMilli(epochMilli)
            .atZone(zoneId)
            .toOffsetDateTime()
    }

    fun toOffsetDateTime(timestamp: Timestamp): OffsetDateTime {
        val instant = timestamp.toInstant()

        return OffsetDateTime.ofInstant(instant, zoneId)
    }

    fun toEpochMilli(offsetDateTime: OffsetDateTime): Long {
        return offsetDateTime.toInstant().toEpochMilli()
    }
}
